source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc

```diff
@@ -103,17 +108,49 @@ void EnvoyQuicServerSession::OnCanWrite() {
 
 void EnvoyQuicServerSession::SetDefaultEncryptionLevel(quic::EncryptionLevel level) {
   quic::QuicServerSessionBase::SetDefaultEncryptionLevel(level);
-  if (level == quic::ENCRYPTION_FORWARD_SECURE) {
-    // This is only reached once, when handshake is done.
-    raiseConnectionEvent(Network::ConnectionEvent::Connected);
+  if (level != quic::ENCRYPTION_FORWARD_SECURE) {
+    return;
   }
+  // This is only reached once, when handshake is done.
+  raiseConnectionEvent(Network::ConnectionEvent::Connected);
+
+  const DetailsWithFilterChain* proof_source_detail =
+      dynamic_cast<const EnvoyQuicCryptoServerStream*>(GetCryptoStream())->proofSourceDetails();
+  if (proof_source_detail == nullptr) {
+    // Unit tests using TestProofSource might not set ProofSource::Details.
+    ENVOY_CONN_LOG(
+        trace,
+        "ProofSource didn't provide ProofSource::Details. No filter chain will be installed.",
+        *this);
+
+    return;
+  }
+  createNetworkFilters(proof_source_detail->filterChain());
 }
 
 bool EnvoyQuicServerSession::hasDataToWrite() { return HasDataToWrite(); }
 
 void EnvoyQuicServerSession::OnOneRttKeysAvailable() {
   quic::QuicServerSessionBase::OnOneRttKeysAvailable();
   raiseConnectionEvent(Network::ConnectionEvent::Connected);
+
+  const DetailsWithFilterChain* details =
+      dynamic_cast<const EnvoyQuicTlsServerHandshaker*>(GetCryptoStream())->proofSourceDetails();
+  if (details == nullptr) {
+    ENVOY_CONN_LOG(
+        trace,
+        "ProofSource didn't provide ProofSource::Details. No filter chain will be installed.",
+        *this);
+    return;
+  }
+  createNetworkFilters(details->filterChain());
```

Comment:
Can we move this into a function and call it from SetDefaultEncryptionLevel and OnOneRttKeysAvailable?